### ISSUE TITLE
[BUG] SFTP File Writer overwrites files despite 'File Exists' set to 'Error

### ISSUE THREAD

**Issue by FlopChecker** [nextgenhealthcare/connect#5829]
**Describe the bug**
The File Writer connector with the method 'sftp' in Mirth Connect is overwriting existing files on the server even though the 'File Exists' option in the destination settings is set to 'Error'.

This might be due to a potential issue inside the class 'SftpConnection'
[source-code](https://github.com/nextgenhealthcare/connect/blob/a874e04ea3a2fe81552cfe7bac973a5bfc9746c2/server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java#L224)

Currently, the method tries to use `client.ls(".").contains(file)` to check if a file exists. However, the `ls` method returns a Vector of ChannelSftp.LsEntry objects, and the `contains` method compares the `String` file with each LsEntry object in the Vector. Since a String can't be equal to an LsEntry object, the contains method always returns false.

**To Reproduce**
Since this issue occurs under normal usage of the SFTP File Writer connector, no special setup steps are required.

1. Set up a channel with an File Writer as the destination, method 'sftp'.
2. In the destination settings, set the 'File Exists' option to 'Error'.
3. Send a message that results in a file being written to the SFTP server.
4. Send another message that would result in a file with the same name being written to the server.
Check the server and see that the original file was overwritten and no error was thrown.

**Expected behavior**
When the 'File Exists' option is set to 'Error', the File Writer should not overwrite existing files; instead it should create an Error or Queued status on the dashboard.

**Actual behavior**
The file is overwritten and no error is thrown.

**Screenshots**
_[Screenshot removed, because it was misleading]_

**Comment by jonbartels** [comment]
Set the file name to an actual file name `*.*` can be a file name but the writer is meant to write to a specific file name. 

Wildcards would be used in a file READER to get sets of files.

**Comment by FlopChecker** [comment]
Sorry about the error in the screenshot I provided, the filename wasn’t related to the bug reported. My apologies for any confusion.
The actual value was the one dropped from the variables (I'm out of office now, I believe '${originalFilename}').

Despite this, I’m still convinced that the main issue — the SFTP File Writer overwriting existing files when ‘File Exists’ is set to ‘Error’ — is a valid bug. I could provide more details and a corrected screenshot (if neccessary at all).

What would be the best way to proceed? Should I add this information to this issue, or would it be better to create a new one? I want to make sure this potential bug gets properly addressed.

**Comment by pacmano1** [comment]
@FlopChecker just tried this, and it does Error for me _if_ the file is set to binary.

If I set it to text, it does not error which I agree is a bug.

**Comment by FlopChecker** [comment]
Thanks @pacmano1 ! That’s interesting... The method ‘exists’ I referred to in the source code seemed to explain the behavior I described, but it doesn’t account for the difference between text and binary. I could be totally off, of course.

We also experimented with the ‘client.ls(”.”).contains(file)’ method from the jcraft library and it consistently returned false.
So, might there be more to this?

**Comment by ReinerNeubert** [comment]
Hello,

I am also involved in this problem. And it's a urgent problem!

This is a productive sample destination connector. This connector overwrites existing SFTP files every time.

![image](https://github.com/nextgenhealthcare/connect/assets/33738170/d6b7d76d-c5b8-4327-8b2f-126189e55591)

Create TMP File also has no effect for this problem.

It should also be clear that append doesn't work either.

This error should also be fixed in previous versions. We can't just upgrade all of our instances to the latest version.

Maybe this problem has existed since Mirth Connect mastered the protocol greater than SMB1.

**The current SFTP function**
```
    @Override
    public boolean exists(String file, String path) {
        try {
            cwd(path);
            return client.ls(".").contains(file);
        } catch (Exception e) {
            return false;
        }
    }
```
**Returns ervery time false!**

We are currently using this Javascript code to check in a previous destination step whether this file exists.
```
try {
   fileList = channel.ls(FileName);
   FilePresent = true;
} catch (err) {
   FilePresent = false;
}
```

jsch is very funny there.
If the file doesn't exist jsch generates an error.

If FilePresent is True we set the status of the Message to Error.

Best Regards
Reiner

**Comment by jonbartels** [comment]
I looked at the source code based on @pacmano1 observation.

I don't see any toggles that would change the file existence check for text vs binary. The flag would be `com.mirth.connect.connectors.file.FileDispatcherProperties#isBinary`

The calls start here:
https://github.com/nextgenhealthcare/connect/blob/e0db5497aabb94c41071e4a098580e660a1d5d8e/server/src/com/mirth/connect/connectors/file/FileDispatcher.java#L168

Then for SFTP:
https://github.com/nextgenhealthcare/connect/blob/e0db5497aabb94c41071e4a098580e660a1d5d8e/server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java#L224-L231

Local file for reference:
https://github.com/nextgenhealthcare/connect/blob/e0db5497aabb94c41071e4a098580e660a1d5d8e/server/src/com/mirth/connect/connectors/file/filesystems/FileConnection.java#L163-L166

@ReinerNeubert showed the same code. I wonder if the problem is with the `cwd`. Is it possible to have file permissions such that an SFTP user could `ls` or write a file but would not have permission to change directories?

**Comment by jonbartels** [comment]
I found this reference: https://github.com/nextgenhealthcare/connect/discussions/5128

There appear to have been other users looking at the same code but I don't exactly see the same conditions https://github.com/search?q=repo%3Anextgenhealthcare%2Fconnect+cwd&type=issues


EDIT - https://github.com/nextgenhealthcare/connect/issues/1852 explains similar conditions with detailed commentary.

**Comment by pacmano1** [comment]
I am root.

**Comment by FlopChecker** [comment]
The problem, from our side, seems not to be the `cwd`.
The issue #3784 seems to exactly describe what I try to highlight. The method `return client.ls(".").contains(file);` will never return true.

I tried to recreate the source code of the current SftpConnection class with the Rhino debugger.

Original SftpConnection code:
 ```
    @Override
    public boolean exists(String file, String path) {
        try {
            cwd(path);
            return client.ls(".").contains(file);
        } catch (Exception e) {
            return false;
        }
    }
```

The results are as follows:
![grafik](https://github.com/nextgenhealthcare/connect/assets/50123209/b4023770-e3da-47c8-9ab1-0543c630a15a)

However, the difference between text vs. binary seems to be a different issue.
We also tried binary vs text on our system, which both led to the same outcome - the file being overwritten without any error.
(Test system is Windows with Buru SFTP for the server)
